Solidus 3 preparation

cpfergus1 opened the pull request:
This prepares the extension to work with current supported versions of Solidus.
This PR satisfies #51 

Discussion points:
Use of conditionals and if parameters to satisfy versioning requirements.
Other possible parameters to use for SoftDelete checks than Solidus Version

jarednorman commented on app/controllers/spree/admin/sale_prices_controller.rb:
```diff
@@ -22,7 +22,7 @@ def create
       def destroy
         @sale_price = Spree::SalePrice.find(params[:id])
-        @sale_price.discard
+        @sale_price.send(SolidusSalePrices::SoftDelete.delete_method)
```
Should probably use `public_send` throughout, just as a matter of safe practices, even if it has no impact to the overall function here.

cpfergus1 replied:
Removed send in favor of #discard. Please let me know if the new implementation works!

jarednorman commented on app/models/spree/sale_price.rb:
```diff
@@ -1,9 +1,18 @@
 module Spree
   class SalePrice < ActiveRecord::Base
-    include Spree::SoftDeletable
+    if self.respond_to?(:acts_as_paranoid)
```
Won't this be true within any application that has paranoia loaded, regardless of the soft deletion method available in Solidus?

cpfergus1 replied:
Has been removed in favor of Solidus version to be more explicit. Please let me know if you have any other ideas in which to based this conditional off of.

kennyadsl commented on app/helpers/solidus_sale_prices/soft_delete.rb:
```diff
@@ -0,0 +1,20 @@
+
+module SolidusSalePrices
+  module SoftDelete
+    def self.delete_method
```
I'm not sure we need this. What if we just always use Discard methods? If I recall correctly, in Solidus 2.10 we had support for both Discard and Paranoia just for backward compatibility, so using discard methods would work exactly the same without emitting deprecation warnings. Did you already try?

cpfergus1 replied:
In Solidus 2.10, Spree::SoftDeletable has not been added yet. This prevents the use of discard methods which is the reason for the conditional methods and I thought it might be cleaner to but some of that logic in a helper

jarednorman replied:
Even without `Spree::SoftDeletable`, discard is still present. I don't see any need to use paranoia.

cpfergus1 replied:
I will explore this further, I was under the impression that the functionality of discard was tied to Spree::SoftDeletable

cpfergus1 replied:
Just started playing with it this morning and found that the specs do fail with just the discard methods implemented. It works and is arguably more concise than the previous implementation if I just force the SoftDeleteable functionality in the price model:
```
    if ::Spree.solidus_gem_version >= Gem::Version.new('2.11')
      include Spree::SoftDeletable
    else
      require 'discard'
      acts_as_paranoid
      include Discard::Model
      self.discard_column = :deleted_at
    end
```
Further, I could throw that into the helper module and place a note to remove when 2.10 is no longer supported as well as create an issue in this repository.

Additionally, without paranoia, the following fails and requires further investigation:
```
  1) Spree::Variant with a valid sale destroys all sale prices when it is destroyed
     Failure/Error:
       expect { variant.discard }
         .to change { Spree::SalePrice.all.size }
         .from(3).to(0)
     
       expected `Spree::SalePrice.all.size` to have changed from 3 to 0, but did not change
     # ./spec/models/variant_spec.rb:144:in `block (3 levels) in <top (required)>'

  2) Spree::Price#destroy when there are sale prices associated to the price destroys all sale prices when it is destroyed
     Failure/Error:
       expect { price.discard }
         .to change { Spree::SalePrice.all.size }
         .from(1).to(0)
     
       expected `Spree::SalePrice.all.size` to have changed from 1 to 0, but did not change
     # ./spec/models/price_spec.rb:92:in `block (4 levels) in <top (required)>'
```
Without Discard::Model, none of the discard methods work.

I will continue to look for other solutions and please let me know if I am steering off in the wrong direction.
Thanks!

kennyadsl reviewed (commented):
Maybe there's the opportunity of removing the conditional code for soft-deletion, I left a comment about it.

cpfergus1 reviewed (commented):
Just started playing with it this morning and found that the specs do fail with just the discard methods implemented. It works and is arguably more concise than the previous implementation if I just force the SoftDeleteable functionality in the price model:
```
    if ::Spree.solidus_gem_version >= Gem::Version.new('2.11')
      include Spree::SoftDeletable
    else
      require 'discard'
      acts_as_paranoid
      include Discard::Model
      self.discard_column = :deleted_at
    end
```
Further, I could throw that into the helper module and place a note to remove when 2.10 is no longer supported as well as create an issue in this repository.

Additionally, without paranoia, the following fails and requires further investigation:
```
  1) Spree::Variant with a valid sale destroys all sale prices when it is destroyed
     Failure/Error:
       expect { variant.discard }
         .to change { Spree::SalePrice.all.size }
         .from(3).to(0)
     
       expected `Spree::SalePrice.all.size` to have changed from 3 to 0, but did not change
     # ./spec/models/variant_spec.rb:144:in `block (3 levels) in <top (required)>'

  2) Spree::Price#destroy when there are sale prices associated to the price destroys all sale prices when it is destroyed
     Failure/Error:
       expect { price.discard }
         .to change { Spree::SalePrice.all.size }
         .from(1).to(0)
     
       expected `Spree::SalePrice.all.size` to have changed from 1 to 0, but did not change
     # ./spec/models/price_spec.rb:92:in `block (4 levels) in <top (required)>'
```
Without Discard::Model, none of the discard methods work.

I will continue to look for other solutions and please let me know if I am steering off in the wrong direction.
Thanks!

kennyadsl commented:
@jarednorman time for a second pass?

jarednorman commented on .byebug_history:
```diff
@@ -0,0 +1,3 @@
+exit
+::Spree.solidus_gem_version
+::Spree.solidus_gem_version > Gem::Version.new('2.10')
```
This shouldn't be committed.

cpfergus1 replied:
Absolutely, Thanks for pointing that out

jarednorman commented on Gemfile:
```diff
@@ -6,11 +6,6 @@ git_source(:github) { |repo| "https://github.com/#{repo}.git" }
-# otherwise it takes forever to resolve them.
-# See https://github.com/bundler/bundler/issues/6677
-gem 'rails', '>0.a'
-
```
Why is this removed?

cpfergus1 replied:
No longer required. There used to be an issue with bundler versions below 2.2 I believe which caused hanging when resolving dependencies for older versions of rails which this line of code rectified. The bug in bundler has been rectified so I was just removing it since I was working on the repo.

jarednorman commented on app/models/spree/sale_price.rb:
```diff
@@ -1,6 +1,11 @@
+    if ::Spree.solidus_gem_version >= Gem::Version.new('2.11')
+      include Spree::SoftDeletable
+    else
+      include SolidusSalePrices::SoftDelete
```
I don't see the point in having this in a concern and including it in only one place. This code is temporary until all supported versions of Solidus have `Spree::SoftDeletable`, so let's make this code easier to delete and just inline it.

cpfergus1 replied:
I see your point on this and looking back at it now, I reimplemented the same functionality of Spree::SoftDeleteable so should we even include the conditional at all? I am thinking of just leaving
```
include SolidusSalePrices::SoftDelete
```
and open an issue to replace that with Spree::SoftDeleteable once 2.10 is no longer supported. What do you think @jarednorman?

jarednorman replied:
I think there's something subtle that I missed on my first review: `SolidusSalePrices::SoftDelete` uses paranoia, but `Spree::SoftDeletable` does not. This means that using the former means you can call `.destroy` on the record and cause it to get soft-deleted, whereas with the latter doing so will cause the record to actually be dropped from the database.

If we care about this difference, I'd say we maintain the current functionality and leave a note in the codebase regarding how we make the breaking change in the future to move to the soft-deletion functionality that comes with Solidus:

```ruby
# The following code enables soft-deletion. In Solidus v2.11+ there is a mixin
# that handles soft-deletion consistently across all Solidus records that need
# it. Once we no longer support v2.10, we can remove the dependency on paranoia
# and replace these lines with:
#
#     include Spree::SoftDeletable
#
# However, this will be a breaking change as it will change the behaviour of
# calling `destroy` on these records.
acts_as_paranoid
include Discard::Model
self.discard_column = :deleted_at
```

jarednorman commented on lib/solidus_sale_prices/engine.rb:
```diff
@@ -4,7 +4,7 @@
 module SolidusSalePrices
   class Engine < Rails::Engine
-    include SolidusSupport::EngineExtensions::Decorators
+    include SolidusSupport::EngineExtensions
```
Typo in commit message: "depreciated" should be "deprecated".

cpfergus1 replied:
Thank you for pointing that out!

jarednorman commented on spec/features/admin/sale_prices_spec.rb:
```diff
@@ -78,14 +78,12 @@
         expect(page).to have_content(32.33)
         expect(page).to have_content(country.name)
-        expect(page).to have_content(variant.prices.last.currency)
+        expect(page).to have_content(variant.prices.first.currency)
```
Unless I'm missing something, this order isn't guaranteed. Either way, the tests should know the currency, so why don't we just make the assertion against the literal?

```suggestion
        expect(page).to have_content("EUR")
```

jarednorman commented on app/helpers/solidus_sale_prices/soft_delete.rb:
```diff
@@ -9,5 +9,6 @@ module SoftDelete
       acts_as_paranoid
       include Discard::Model
       self.discard_column = :deleted_at
+    end
```
I'd appreciate a pass to clean up the commit sequence, so changes like this are in the right commits.

cpfergus1 replied:
Done!

jarednorman commented on spec/models/variant_spec.rb:
```diff
@@ -24,18 +22,15 @@
-
     variant.prices.each do |p|
-      variant.put_on_sale 10.95, { currencies: [ p.currency ] }
+     variant.put_on_sale 10.95, { currencies: [ p.currency ] }
```
This block should be indented two spaces, not one.

jarednorman commented on spec/models/sale_price_spec.rb:
```diff
@@ -39,7 +39,7 @@
 
     it 'sets the end time' do
-      expect(sale_price.end_at).to be_within(1.second).of(Time.now)
+      expect(sale_price.end_at).to be_within(2.second).of(Time.now)
```
I don't actually care if you fix, but fwiw in the commit message "accomidate" should be "accommodate".

cpfergus1 replied:
Thank you for catching that, I will get it rectified.

jarednorman reviewed (changes requested):
Overall approach seems good to me. This could use some rebase love to get rid of some churn and things committed in the wrong commits to keep the history sensible.

cpfergus1 commented:
Thank you for your corrections @kennyadsl and @jarednorman. All current topics have been remedied and pushed. I also removed two older commits that were obsolete and cleaned up the history a little bit to be more explicit. Please let me know if I left anything out!
